🚀 1단계 - 문자열 덧셈 계산기

MoonHKLee opened the pull request:
리뷰 잘 부탁드립니다.

역시 이전 코드와 섞여서 커밋이 더럽네요...

wotjd243 committed:
build: upgrade to spring boot 3.2.5 and enable kotlin

MoonHKLee committed:
test(racingcar) : 레이싱카 이름 관련 테스트 작성.

MoonHKLee committed:
feature(racingcar) : 레이싱카 이름 관련 기능 구현.

MoonHKLee committed:
refactor(racingcar) : 레이싱카 변수 이름 변경.

MoonHKLee committed:
feature(racingcar) : 레이싱카 요구사항 구현 및 테스트코드 작성 완료.

MoonHKLee committed:
Merge branch 'moonhklee' into moonhklee

MoonHKLee committed:
Merge remote-tracking branch 'upstream/moonhklee' into moonhklee

MoonHKLee committed:
chore : 이전 작업 삭제.

MoonHKLee committed:
docs : 요구사항 작성

MoonHKLee committed:
test : 요구사항 관련 테스트코드 작성.

MoonHKLee committed:
feature(string-calculator) : 문자열 계산기 구현코드 작성 완료.

MoonHKLee committed:
refactor(string-calculator) : final 선언

MoonHKLee committed:
refactor(string-calculator) : Number class access modifier 수정. default -> public

MoonHKLee committed:
test(string-calculator) : 예외 메시지를 영어로 변경하고 예외 처리 테스트 확장

예외 메시지가 한글에서 영어로 변경되었습니다. 이 변화는 문자열 계산기의 내부 로직과 예외 처리 과정에 영향을 주었습니다. 더불어 예외 처리 테스트도 상세화하였습니다.

MoonHKLee committed:
test(number) : 숫자 관련 테스트 코드 추가

새로운 파일 NumberTest.java를 생성하여 숫자 생성, 문자열을 숫자로의 변환 실패 와 음수 예외 처리에 대한 테스트 코드를 작성하였습니다. 이를 통해 코드의 안정성을 높였습니다.

MoonHKLee committed:
fix(number) : 오타 수정

'Number' 클래스와 'StringCalculatorTest' 클래스에서 예외 메시지의 오타를 수정했습니다. "Number must me positive."를 "Number must be positive."로 변경하였습니다. 이로인해 오류 메시지가 더 정확하게 사용자에게 전달될 수 있습니다.

MoonHKLee committed:
test(calculator) : Calculator 테스트 코드 추가

새로운 CalculatorTest.java 파일을 추가하였습니다. 이 테스트 코드는 Calculator의 sum 메서드가 수의 스트림을 올바르게 합산하는지 확인합니다. Assert 구문을 통해 결과값이 예상치에 맞게 동작하는지 검증합니다.

MoonHKLee committed:
fix(number) : Number 클래스에 equals 및 hashCode 구현

Number 클래스에 Java의 내장 함수 Objects.hashCode()를 사용하여 hashCode를 오버라이드하고, 적절한 equals 메서드를 구현하였습니다. 이와 동시에, 새로운 파일인 NumberMapperTest.java를 추가하여 NumberMapper 클래스의 테스트 케이스를 작성하였습니다. 이로써 숫자 관련 연산을 더욱 안정적으로 수행할 수 있습니다.

MoonHKLee committed:
test(text-input-handler) : 테스트 메소드 이름 변경 및 TextInputHandlerTest 추가

StringCalculatorTest의 테스트 메소드 이름을 좀 더 명확하게 변경하였습니다. 추가적으로, TextInputHandlerTest를 새로 생성하여, 문자열 토큰화 및 문자열 null 또는 비어있는 경우에 대한 테스트를 추가로 구현하였습니다. 이제 각 기능별로 꼼꼼하게 테스트할 수 있습니다.

MoonHKLee committed:
chore : 패키지 calculator2에서 calculator로 이름 변경

파일들이 calculator2 패키지에서 calculator 패키지로 이동되었습니다. 이에 따라 모든 파일들의 패키지 이름이 calculator2에서 calculator로 변경되었습니다. 이 변동 사항은 이 프로젝트의 정리와 관리를 향상시키는데 도움이 될 것입니다.

MoonHKLee committed:
style : 코드 정리 및 임포트 구문 재정렬

이 커밋은 코드 내의 공백을 제거하며, 여러 파일의 임포트 구문의 순서를 변경합니다. 또한 테스트 코드에서 문자열 배열의 포맷팅을 간결하게 하였습니다.

kyucumber commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,10 @@
+
+public class Calculator {
+
+    public int sum(Stream<Number> numbers) {
```
Stream을 매개 변수나 리턴 타입으로 쓰는건 지양해주세요. `List`와 같은 컬렉션을 바로 스트림으로 바꿀 수 있으므로 어떠한 이점도 없으며 이미 다른곳에서 소비된 스트림을 받아 예외가 발생할 수 있는 여지도 있습니다.

MoonHKLee replied:
왜 stream으로 반환했는지 모르겠네요.. 하하

kyucumber commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,10 @@
+public class Calculator {
+
+    public int sum(Stream<Number> numbers) {
+        return numbers.mapToInt(Number::getNumber).sum();
```
mapToInt에서 getNumber를 하나씩 받아서 사용하고 있네요.  Numbers라는 일급 컬렉션을 만들고 sum이란 메서드를 제공하는 방법도 있을 것 같아요.

MoonHKLee replied:
일급컬렉션 적용하는게 코드가 훨씬 줄어드네요

kyucumber commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,40 @@
+
+    private Number(int number) {
+        if (number < 0) {
+            throw new RuntimeException("Number must be positive.");
```
RuntimeException이 아닌 RuntimeException의 구현체인 IllegalArgument, State 등을 사용하시는건 어떨까요?

MoonHKLee replied:
RuntimeException 잡으라고해서 그대로 Runtime으로 잡아버렸네요

kyucumber commented on src/main/java/calculator/TextInputHandler.java:
```diff
@@ -0,0 +1,23 @@
+    private static final String DEFAULT_DELIMITERS = ",|:";
+
+    public String[] tokenize(String text) {
+        Matcher matcher = Pattern.compile(CUSTOM_DELIMITER_SYNTAX).matcher(text);
```
지금 정도에서는 부하가 없겠지만 Pattern은 생성 비용이 큰 객체입니다. API 서버라고 가정했을 때 매 요청마다 Pattern이란 객체를 생성하는건 부담이 됩니다. 정적 값으로 정의해서 활용하는건 어떨까요?

MoonHKLee replied:
static으로 처리할수 있는 부분은 일단 최대한 캐싱해야겠네요.
불필요한 객체 생성 없애도록 하겠습니다.

kyucumber commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,94 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+class StringCalculatorTest {
```
테스트를 잘 작성해주셨네요 👍

MoonHKLee replied:
감사합니다.

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드려요.

MoonHKLee committed:
refactor(string-calculator) : Calculator과 NumberMapper 제거 및 기능 리팩토링

Calculator, NumberMapper 클래스들이 제거되고, 이들의 기능은 Numbers 클래스로 이동되었습니다. 더욱이, CalculatorTest의 이름도 NumbersTest로 변경되었습니다. 예외의 경우, RuntimeException 대신 IllegalArgumentException을 사용하여 더 세밀한 예외 처리를 보장하도록 수정하였습니다.

kyucumber reviewed (approved):
코멘트 잘 반영해주셨습니다.
다음 단계 진행을 위해 머지할게요
